### ISSUE TITLE
Build fails with --enable-g=most

### ISSUE THREAD

**Issue by wgropp** [pmodels/mpich#3731]
I'm trying to build mpich and my build fails.  I've cloned a new copy to ensure that I have a clean source, and I have followed the instructions for a fresh checkout.

My configure line, appropriate for development, is 

> $builddir/configure --prefix=/Users/gropp/installs/mpich2-dbg \
        --enable-error-checking=runtime \
        --enable-g=most \
        --enable-threads=runtime \
        --with-pm=gforker:hydra \
        --with-device=ch3

The build fails; see the attached make.log.  It appears to be some conflict with the memory tracing routines and some code that is using raw malloc/calloc/free .
[make.log](https://github.com/pmodels/mpich/files/3060347/make.log)

**Comment by hzhou** [comment]
I remember seeing the same error, it can be reproduced on Linux. Oddly, `--enable-g=all` will work -- and that is everyone else has been using.

`--enable-g=all` also sets `perform_memarena=yes` in addition to all the settings of `--enable-g=most`.

**Comment by hzhou** [comment]
This is due to inconsistency between `configure.ac` (in MPICH) and `src/mpl/configure.ac`. `--enable-g=most` is handled by `mpich` `configure`, but ignored by `mpl` `configure`. 

When "memory tracing" is on, `include/mpir_mem.h` defines `malloc` etc to `Error` while `mpl_trmem.h` sets `MPL_malloc` to `MPL_trmalloc`. We can see how `mpir_mem.h` is on, but `mpl_trmem.h` is off will essentially cause error in all these functions.

**Comment by wgropp** [comment]
Most should be consistent.

This should also have been caught by the test that picks a different collection of configure options at random and does a configure/make to check for problems with the less used options.  Given the number of options just for ch4, using this test is essential.  Did it not catch this problem?

Bill

William Gropp
Director and Chief Scientist, NCSA
Thomas M. Siebel Chair in Computer Science
University of Illinois Urbana-Champaign






On Apr 9, 2019, at 11:34 PM, Hui Zhou <notifications@github.com<mailto:notifications@github.com>> wrote:


This is due to inconsistency between configure.ac (in MPICH) and src/mpl/configure.ac. --enable-g=most is handled by mpich configure, but ignored by mpl configure.

When "memory tracing" is on, include/mpir_mem.h defines malloc etc to Error while mpl_trmem.h sets MPL_malloc to MPL_trmalloc. We can see how mpir_mem.h is on, but mpl_trmem.h is off will essentially cause error in all these functions.

**Comment by raffenet** [comment]
We have a random config test in Jenkins, but I suspect it may not actually be random...

See random column:
https://jenkins.mpich.org/job/mpich-master-ch3-ubuntu/

**Comment by hzhou** [comment]
@raffenet TIL that we have a random test :)

Following are the lists (from `test-worker.sh`):
```
enable_array=(
        'error-checking;no;runtime;all'
        'error-messages;all;generic;class;none'
        'timer-type;linux86_cycle;clock_gettime;gettimeofday'
        'timing;none;all;runtime;log;log_detailed'
        'g;none;all;handle;dbg;log;meminit;handlealloc;instr;mem;mutex;mutexnesting'
        'fast;O0;O1;O2;O3;ndebug;all;yes;none'
        'fortran'
        'cxx'
        'romio'
        'check-compiler-flags'
        'strict;c99;posix'
        'debuginfo'
        'weak-symbols;no;yes'
        'threads;single;multiple;runtime'
        'thread-cs;global'
        'refcount;lock-free;none'
        'mutex-timing'
        'handle-allocation;tls;mutex'
        'multi-aliases'
        'predefined-refcount'
        'alloca'
        'yield;sched_yield;select'
        'runtimevalues'
    )
    with_array=(
        'logging;none'
        'pmi;simple'
        'pm;gforker'
        'namepublisher;no;file'
        'device;ch3;ch3:sock'
        'shared-memory;sysv'
    )
```
`g=most` happen to be not in it :(

**Comment by hzhou** [comment]
I checked the `config.log` of several past "random" tests, none of them have chosen any options -- the script must be broken.

**Comment by raffenet** [comment]
😩 we should fix that, too.

**Comment by raffenet** [comment]
Build issue is fixed in #3735. Will create separate issue for Jenkins random tests.
